packages/sls-env/src/dependencies.ts

```diff
@@ -70,22 +95,33 @@ const resolveEventDependencies = <D, P, E, C>(
   context: C
 ): AppDependencyConverter<D> => {
   const updatedDependencies = Object.entries(dependencies).reduce((acc, [dependencyName, dependency]) => {
-    if (dependency?.type === 'EventDependency') {
-      const key = dependency.getKey(payload, event, context)
-      const eventDependency = dependency.dependencies[key]
-      if (!eventDependency) {
-        logger.warn(`could not extract event based dependency for ${dependencyName}`)
-        throw new EventDependencyError(`No event based dependency found for '${dependencyName}'`)
-      } else {
-        logger.debug(`replacing '${dependencyName}' with event based dependency`)
-        return { ...acc, [dependencyName]: eventDependency }
-      }
+    switch (dependency.type) {
+      case DependencyType.EventDependency:
+        const key = dependency.getKey(payload, event, context)
+        const eventDependency = dependency.dependencies[key]
+        if (!eventDependency) {
+          logger.warn(`could not extract event based dependency for ${dependencyName}`)
+          throw new EventDependencyError(`No event based dependency found for '${dependencyName}'`)
+        } else {
+          logger.debug(`replacing '${dependencyName}' with event based dependency`)
+          return { ...acc, [dependencyName]: eventDependency }
+        }
+      case DependencyType.DependencyFactory:
+        logger.debug(`replacing '${dependencyName}' with factory based dependency`)
+        return { ...acc, [dependencyName]: dependency.dependencyFactory({ payload, event, context }) }
```

Comment:
when would you use one over the other, we have to document that and how would you prevent dependency factory to create dependency each time and prevent from being cached?

Reply:
I think both methods try to solve the same problem in different ways.

the `eventDependency` method forces the user to instantiate all dependencies ahead of time.
and also provide a `getKey` function which internally within the sls-env package will get used to pick the right dependency from the `dependencies` object.
```
myDependency: {
          type: 'EventDependency',
          dependencies: {
            'variant1': {//dependency variant 1},
            'variant2': {//dependency variant 1},
            'variant3': {//dependency variant 1},
              // ...
          }},
          getKey: (params) => {//return 'variant1', 'variant2', 'variant3'....}
        }
```
If there are many variations for the dependencies all of them have to be created on lambda init.
This would use a lot more memory than needed
If we are trying to build an open-source project I think the library should be as flexible as possible and let the user decide how to resolve their dependency.

the `eventDependencyFactory` is following the same idea however it is allowing the user decide how they want to resolve their dependency once they have access to the event/payload/context.

It will be up to the factory itself to cache the resolved dependencies based on the event/payload/context
See an example of this in `createS3EventDependantAppClientFactory` in https://github.com/boohoo-com/debenhams/blob/56bcf524257284a9ab7d87eeaecefd8bd08d6a03/packages/commercetools/src/client/event/create-client-from-s3-eventbridge-event.ts

I have created this alongside `eventDependency` to allow backwards compatibility and not having to rewrite already working solutions with `eventDependency`

However if we decide the dependency factory to be a better method I could create a ticket to update those usages.

Reply:
I have added a wrapper around the `resolveDependency` function so that we return a new EventDependencyError() here when the factory throws.